### PR TITLE
Clarify boolean conditional evaluation

### PR DESCRIPTION
**Pull request by leafpetersen**
Clarify the behavior of boolean conditional expressions in strong and weak mode.

cc @nshahan @sigmundch @rakudrama @alexmarkov @lrhn @eernstg @munificent

### CODE REVIEW

File: accepted/future-releases/nnbd/feature-specification.md [hunk 1]
```diff
@@ -887,6 +890,42 @@ was marked `late`.  Note that this is a change from pre-NNBD semantics in that:
...
+assignable to `bool` is unchanged from pre null-safe Dart.  The change in
+assignability means that the static type of the condition may only be `dynamic`,
+`Never`, or `bool`.  In full null-safe Dart, an expression of type `Never` will
+always diverge and an expression of type `bool` will never evaluate to a value
+other than `true` or `false`, and hence no conversion is required in these
+cases.  A conditional expression of type `dynamic` may evaluated to any value,
+and hence must be implicitly downcast to `bool`, after which no further check is
+required.
+
+In mixed mode (weak mode evaluation) Dart, values of type `Never` and `bool` may
```
Comment (alexmarkov) [inline accepted/future-releases/nnbd/feature-specification.md:906]: > In mixed mode (weak mode evaluation) Dart

It doesn't matter if it is mixed or not. Maybe simply 'In weak checking mode...'

Comment (leafpetersen) [reply accepted/future-releases/nnbd/feature-specification.md]: Done.

**Comment by leafpetersen** [comment]
Landing this.  I'm working on tests to accompany, I will file issues for any failures.

**Comment by leafpetersen** [comment]
Tests out for review here: https://dart-review.googlesource.com/c/sdk/+/142782

**Pull request merged**